src/garage/torch/algos/vpg.py

```diff
@@ -128,92 +134,168 @@ def train_once(self, itr, paths):
 
-        with torch.no_grad():
-            loss_after = self._compute_loss(itr, obs, actions, rewards, valids,
-                                            baselines)
-            kl = self._compute_kl_constraint(obs)
-            policy_entropy = self._compute_policy_entropy(obs)
+        return loss
 
-        average_returns = log_performance(itr,
-                                          TrajectoryBatch.from_trajectory_list(
-                                              self.env_spec, paths),
-                                          discount=self.discount)
+    def _compute_loss(self, obs, actions, rewards, valids, baselines):
+        """Compute mean value of loss.
 
-        with tabular.prefix(self.policy.name):
-            tabular.record('LossBefore', loss.item())
-            tabular.record('LossAfter', loss_after.item())
-            tabular.record('dLoss', loss.item() - loss_after.item())
-            tabular.record('KLBefore', kl_before.item())
-            tabular.record('KL', kl.item())
-            tabular.record('Entropy', policy_entropy.mean().item())
+        Args:
+            obs (torch.Tensor): Observation from the environment.
+                with shape :math:`(N, max_path_length, observation_dim)`.
+            actions (torch.Tensor): Actions fed to the environment.
+                with shape :math:`(N, max_path_length, action_dim)`.
+            rewards (torch.Tensor): Acquired rewards.
+                with shape :math:`(N, max_path_length)`.
+            valids (list[int]): Numbers of valid steps in each paths.
+            baselines (torch.Tensor): Value function estimation at each step
+                with shape :math:`(N, max_path_length)`.
 
-        self.baseline.fit(paths)
-        return np.mean(average_returns)
+        Returns:
+            torch.Tensor: Calculated negative mean scalar value of objective
 
-    def _compute_loss(self, itr, obs, actions, rewards, valids, baselines):
+        """
+        obs_flat = torch.cat(filter_valids(obs, valids))
+        actions_flat = torch.cat(filter_valids(actions, valids))
+        rewards_flat = torch.cat(filter_valids(rewards, valids))
+        advantages_flat = self._compute_advantage(rewards, valids, baselines)
+
+        return self._compute_loss_with_adv(obs_flat, actions_flat,
+                                           rewards_flat, advantages_flat)
+
+    def _compute_loss_with_adv(self, obs, actions, rewards, advantages):
         """Compute mean value of loss.
 
         Args:
-            itr (int): Iteration number.
             obs (torch.Tensor): Observation from the environment.
-            actions (torch.Tensor): Predicted action.
-            rewards (torch.Tensor): Feedback from the environment.
-            valids (list[int]): Array of length of the valid values.
-            baselines (torch.Tensor): Value function estimation at each step.
+                with shape :math:`(T, observation_dim)`.
```

Comment:
i think you mean `(N \dot [T], O*)` here, and etc below? see the terminology in 

https://github.com/rlworkgroup/garage/blob/master/src/garage/_dtypes.py